roles/setup_lvm_vg/action_plugins/validate_storage.py

```diff
@@ -0,0 +1,160 @@
+'''
+
+# Note: to add mode checks/validations, simply add a top-level function whose
+# name starts with `check_`. The function will receive a `task_vars` dictionary
+# as defined by Ansible.
+# Within a `check_*` function, use `assert` to validate values found in
+# `task_vars`, or raise `AssertionError` explicitly. A single check can, of
+# course, contain multiple assertions.
+# Alternatively, for checks which can result in multiple errors, a check can
+# return a list of (or yield) error messages.
+
+from ansible.plugins.action import ActionBase
+
+import inspect
+import sys
+
+
+def is_device_present(device, ansible_devices):
+    '''
+    Check that the device passed in parameters exists
+
+    The device can be either a raw device or a partition but nothing else.
+
+    Params:
+        device (string): The name of the device
+            i.e: 'sdb' or 'sdb1'
+        ansible_devices (dict): The dictionary of devices gather by the 'setup'
+            ansible module
+
+    Returns:
+        bool: True if the device is in ansible_device.
+            False otherwise
+    '''
+
+    # if the device is a raw device, check its presence
+    if device in ansible_devices:
+        return True
+
+    # otherwise, check if it's a partition in other devices
+    for ansible_device, device_attr in ansible_devices.items():
+        if device in device_attr['partitions']:
+            return True
+
+    return False
+
+
+def check_devices_presence(hostvars):
+    '''
+    Check that the devices specified in metalk8S_lvm_drives_<vg name>
+    exists on the server
+
+    Params:
+        hostvars (dict): The dictionary 'hostvars' from 'setup' ansible module
+            for a specific host
+
+    Raises:
+        AssertionError: if the device is not present on the host
+    '''
+
+    for lvm_vg in \
+        hostvars.get(
+            'metalk8s_lvm_all_vgs', {}).keys():
+        mk8s_vg_drives_var = 'metalk8s_lvm_drives_' + lvm_vg
+        for device in hostvars.get(mk8s_vg_drives_var):
+            device_name = device.split('/')[-1]
+            assert is_device_present(
+                device_name, hostvars['ansible_devices']), \
+                "The device {} is not present".format(device)
+
+
+class ActionModule(ActionBase):
+    '''
+    Check storage configuration against the current setup for MetalK8s
+    '''
+
+    def run(self, tmp=None, task_vars=None):
+        if task_vars is None:
+            task_vars = dict()
+
+        result = super(ActionModule, self).run(tmp, task_vars)
+        del tmp  # tmp no longer has any effect
+
+        def collect_checks():
+            for (name, obj) in inspect.getmembers(sys.modules[__name__]):
+                if name.startswith('check_') and inspect.isfunction(obj):
+                    yield (name, obj)
+
+        errors = []
+        failed = False
+
+        for (name, check) in collect_checks():
+            for host in task_vars['hostvars'].keys():
```

Comment:
We could also restrict to 'kube-nodes' instead of all nodes

Reply:
Indeed. Although it works because the variable metalk8s_lvm_all_vgs will be only defined on hosts with storage, it seems more logical to restrain the scope of the hosts tested.